stdlib/fnmatch.pyi

```diff
@@ -1,9 +1,10 @@
 from collections.abc import Iterable
-from typing import AnyStr
+from pathlib import Path
+from typing import AnyStr, Union
 
 __all__ = ["filter", "fnmatch", "fnmatchcase", "translate"]
 
-def fnmatch(name: AnyStr, pat: AnyStr) -> bool: ...
-def fnmatchcase(name: AnyStr, pat: AnyStr) -> bool: ...
-def filter(names: Iterable[AnyStr], pat: AnyStr) -> list[AnyStr]: ...
+def fnmatch(name: Union[AnyStr, Path], pat: AnyStr) -> bool: ...
+def fnmatchcase(name: Union[AnyStr, Path], pat: AnyStr) -> bool: ...
+def filter(names: Iterable[Union[AnyStr, Path]], pat: AnyStr) -> list[AnyStr]: ...
```

Comment:
If this PR _is_ accepted, there's two issues here. The first is that a `pathlib.Path` object is a wrapper around a string path (not a bytes path), so if you use a bytes object for the second argument, this will fail:

```pycon
>>> import fnmatch, pathlib
>>> fnmatch.fnmatch(pathlib.Path("foo"), b"bar")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alexw/.pyenv/versions/3.12.4/lib/python3.12/fnmatch.py", line 36, in fnmatch
    return fnmatchcase(name, pat)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.12.4/lib/python3.12/fnmatch.py", line 71, in fnmatchcase
    return match(name) is not None
           ^^^^^^^^^^^
TypeError: cannot use a bytes pattern on a string-like object
```

The type annotations need to express that either both arguments have to be string paths, or they both have to be bytes paths.

The second issue is that any `os.PathLike` object will do, not just `pathlib.Path` objects.

You can solve both issues by using our generic type alias we define here:

https://github.com/python/typeshed/blob/bed832da8a30c3d8a8a03aead449f24f2631c3ed/stdlib/_typeshed/__init__.pyi#L173-L176

```suggestion
def fnmatch(name: GenericPath[AnyStr], pat: AnyStr) -> bool: ...
def fnmatchcase(name: GenericPath[AnyStr], pat: AnyStr) -> bool: ...
def filter(names: Iterable[GenericPath[AnyStr]], pat: AnyStr) -> list[AnyStr]: ...
```

Reply:
Thanks for the advice! I'm trying to wrap my head around this solution. The reason that the previous code prevented mixing `str` and `bytes` is because `typing.AnyStr` is a _constrained_ type variable, meaning it's only allowed to have a single value in each function call at a time. But I'm still confused why we're using `GenericPath[AnyStr]` instead of `GenericPath`. Does the definition of the `TypeAlias` already include `AnyStr`?

Reply:
> The reason that the previous code prevented mixing `str` and `bytes` is because `typing.AnyStr` is a _constrained_ type variable, meaning it's only allowed to have a single value in each function call at a time.

Correct

> Does the definition of the `TypeAlias` already include `AnyStr`?

Yup! The type alias is defined like this:

```py
 GenericPath: TypeAlias = AnyStr | PathLike[AnyStr] 
```

Because `AnyStr` in this type alias is a `TypeVar`, that leads to the type alias being implicitly generic around the `AnyStr` typevar. If you think type aliases being implicitly generic if they contain a typevar on the right-hand-side, you're not alone! This has been solved as part of PEP-695. On Python 3.12+, you can create explicitly generic type aliases like this:

```py
type GenericPath[AnyStr: (str, bytes)] = AnyStr | PathLike[AnyStr]
```

Unfortunately, that's syntax that we can't use yet in typeshed, however (we'll be able to do so once we drop support for Python 3.11, but that's a long way off).

Reply:
It seems I still have a lot to learn about type hints, I barely understand what any of this means 😆

Reply:
When a type alias uses a typevar, in this case `AnyStr`, it means that the type alias should be used with `[]` after it. Here is another example. (This used to be in tkinter stubs, but I deleted it at some point because it was unnecessary.)

```python
_T = TypeVar("_T")
_TkinterSequence: TypeAlias = list[_T] | tuple[_T, ...]
```

This means that `_TkinterSequence[str]` means "list or tuple of strings". If you write `_TkinterSequence`, type checkers will likely interpret it as `_TkinterSequence[Any]`, so `list[Any] | tuple[Any, ...]`, which basically means any list or tuple.

Similarly, when you write `GenericPath[Foo]`, it is equivalent to `Foo | PathLike[Foo]`. For example, `GenericPath[str]` means `str | PathLike[str]`. (We would usually use `StrPath` for that, but `GenericPath[str]` would be correct too.) So if you write `GenericPath`, it means `PathLike[Any] | Any`, which accepts anything.

As Alex points out, this is weird, and new typing syntax makes it slightly less weird.

If you write `GenericPath[AnyStr]`, it does what you want, and you end up with definitions like this:

```python
def fnmatch(name: AnyStr | PathLike[AnyStr], pat: AnyStr) -> bool: ...
```

If `AnyStr` resolves to `str`, this becomes `fnmatch(name: str | PathLike[str], pat: str)`, where `PathLike[str]` basically means `pathlib.Path` and similar objects.

If `AnyStr` resolves to `bytes`, this becomes `fnmatch(name: bytes | PathLike[bytes], pat: bytes)`.